pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -498,30 +509,124 @@ func (q *queue) worker(ctx context.Context, f osqueue.RunFunc) {
-			//
-			// loop:
-			// - extend lease for shadow partition
-			// - iterate through backlogs
-			//   + add item to function partition
-			//
+	// acquire lease for shadow partition
+	leaseID, err := duration(ctx, q.primaryQueueShard.Name, "shadow_partition_lease", q.clock.Now(), func(ctx context.Context) (*ulid.ULID, error) {
+		leaseID, err := q.ShadowPartitionLease(ctx, shadowPart, ShadowPartitionLeaseDuration)
+		return leaseID, err
+	})
+	if err != nil {
+		if errors.Is(err, ErrShadowPartitionAlreadyLeased) {
+			// contention
+			return nil
+		}
+	}
+
+	if leaseID == nil {
+		return fmt.Errorf("missing shadow partition leaseID")
+	}
+
+	extendLeaseCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// keep extending lease while we're processing
+	go func() {
+		for {
+			select {
+			case <-extendLeaseCtx.Done():
+				return
+			case <-time.Tick(ShadowPartitionLeaseDuration / 2):
+				leaseID, err = q.ShadowPartitionExtendLease(ctx, shadowPart, *leaseID, ShadowPartitionLeaseDuration)
+				if err != nil {
+					if errors.Is(err, ErrShadowPartitionAlreadyLeased) || errors.Is(err, ErrShadowPartitionLeaseNotFound) {
+						// contention
+						return
+					}
+					return
+				}
+			}
+		}
+	}()
+
+	// Check if shadow partition cannot be processed (paused/refill disabled, etc.)
+	if shadowPart.PauseRefill {
+		q.removeShadowContinue(ctx, shadowPart, false)
+
+		forceRequeueAt := q.clock.Now().Add(ShadowPartitionRefillPausedRequeueExtension)
+		err = q.ShadowPartitionRequeue(ctx, shadowPart, *leaseID, &forceRequeueAt)
+		if err != nil {
+			return fmt.Errorf("could not requeue shadow partition: %w", err)
+		}
+
+		return nil
+	}
+
+	until := q.clock.Now()
+	limit := ShadowPartitionPeekMaxBacklogs
+	backlogs, err := q.ShadowPartitionPeek(ctx, shadowPart, until, limit)
+	if err != nil {
+		return fmt.Errorf("could not peek backlogs for shadow partition: %w", err)
+	}
+
+	// Sequentially refill backlogs
+	for _, backlog := range backlogs {
+		err := q.BacklogRefill(ctx, backlog, shadowPart)
+		if err != nil {
+			// TODO Handle capacity reached and exit
+			isCapacityReached := false
+			if isCapacityReached {
+				q.removeShadowContinue(ctx, shadowPart, false)
+
+				forceRequeueAt := q.clock.Now().Add(ShadowPartitionRefillCapacityReachedRequeueExtension)
+				err = q.ShadowPartitionRequeue(ctx, shadowPart, *leaseID, &forceRequeueAt)
+				if err != nil {
+					return fmt.Errorf("could not requeue shadow partition: %w", err)
+				}
+
+				return nil
+			}
+		}
+	}
 
-			// TODO: probably good to wrap this in a closure with the shadow partition logic
-			// so it can just do a defer
-			metrics.ActiveShadowScanerCount(ctx, -1, metrics.CounterOpt{PkgName: pkgName})
+	hasMoreBacklogs := false
+	if !hasMoreBacklogs {
+		// No more backlogs right now, we can continue the scan loop until new items are added
+		q.removeShadowContinue(ctx, shadowPart, false)
+
+		err = q.ShadowPartitionRequeue(ctx, shadowPart, *leaseID, nil)
```

Comment:
I wonder if the shadow partition should just be deleted if there are no more backlogs in it instead of requeueing